Sources/SwiftProtobuf/TextDecoder.swift

```diff
@@ -0,0 +1,211 @@
+        try scanner.skipRequired(token: .colon)
+        if try scanner.skipOptional(token: .beginArray) {
+            var firstItem = true
+            while true {
+                if try scanner.skipOptional(token: .endArray) {
+                    return
+                }
+                if firstItem {
+                    firstItem = false
+                } else {
+                    try scanner.skipRequired(token: .comma)
+                }
+                try S.setFromText(scanner: scanner, value: &value)
+            }
+        } else {
+            try S.setFromText(scanner: scanner, value: &value)
+        }
+    }
+
+    mutating func decodePackedField<S: FieldType>(fieldType: S.Type, value: inout [S.BaseType]) throws {
+        try decodeRepeatedField(fieldType: fieldType, value: &value)
+    }
+
+    mutating func decodeSingularMessageField<M: Message>(fieldType: M.Type, value: inout M?) throws {
+        _ = try scanner.skipOptional(token: .colon)
+        try M.setFromText(scanner: scanner, value: &value)
+    }
+
+    mutating func decodeRepeatedMessageField<M: Message>(fieldType: M.Type, value: inout [M]) throws {
+        _ = try scanner.skipOptional(token: .colon)
+        if try scanner.skipOptional(token: .beginArray) {
+            var firstItem = true
+            while true {
+                if try scanner.skipOptional(token: .endArray) {
+                    return
+                }
+                if firstItem {
+                    firstItem = false
+                } else {
+                    try scanner.skipRequired(token: .comma)
+                }
+                try M.setFromText(scanner: scanner, value: &value)
+            }
+        } else {
+            try M.setFromText(scanner: scanner, value: &value)
+        }
+    }
+
+    mutating func decodeSingularGroupField<G: Message>(fieldType: G.Type, value: inout G?) throws {
+        try decodeSingularMessageField(fieldType: fieldType, value: &value)
+    }
+
+    mutating func decodeRepeatedGroupField<G: Message>(fieldType: G.Type, value: inout [G]) throws {
+        try decodeRepeatedMessageField(fieldType: fieldType, value: &value)
+    }
+    
+    private func decodeMapEntry<KeyType: MapKeyType, ValueType: MapValueType>(mapType: ProtobufMap<KeyType, ValueType>.Type, keyField: inout KeyType.BaseType?, valueField: inout ValueType.BaseType?) throws where KeyType.BaseType: Hashable {
+        let terminator = try scanner.readObjectStart()
+        while let token = try scanner.next() {
+            if token == terminator {
+                return
+            }
+            switch token {
+            case .identifier("key"):
+                _ = try scanner.skipRequired(token: .colon)
+                try KeyType.setFromText(scanner: scanner, value: &keyField)
+            case .identifier("value"):
+                // Awkward:  If the value is message-typed, the colon is optional,
+                // otherwise, it's required.
```

Comment:
I think so, but I'm not entirely sure.  Given how organically Text format has evolved, it's not entirely clear what "correct" means.  (There are good reasons that no-one has tried to document this format.)  Regardless, I think this is pretty safe:  I'm pretty sure this does not introduce any ambiguity (there's no possibility of mis-parsing correct data), and the C++ implementation generally errs on the side of being overly accepting.

(I'm actually more worried that I've misunderstood somewhere and the ':' is really optional for all field types.  Fortunately, that would be a very easy fix.)